src/OrchardCore/OrchardCore.Deployment.Abstractions/DeploymentSourceBase.cs

```diff
@@ -0,0 +1,26 @@
+namespace OrchardCore.Deployment;
+
+/// <summary>
+/// Represents a base class for deployment source.
+/// </summary>
+/// <typeparam name="TDeploymentStep">The deployment step type.</typeparam>
+public abstract class DeploymentSourceBase<TDeploymentStep>
+    : IDeploymentSource<TDeploymentStep> where TDeploymentStep : DeploymentStep
+{
+    public TDeploymentStep DeploymentStep { get; private set; }
+
+    /// <inheritdoc/>
+    public async Task ProcessDeploymentStepAsync(DeploymentStep step, DeploymentPlanResult result)
+    {
+        if (step is not TDeploymentStep)
+        {
+            return;
+        }
+
+        DeploymentStep = step as TDeploymentStep;
+
+        await ProcessDeploymentStepAsync(result);
+    }
```

Comment:
```suggestion
    /// <inheritdoc/>
    public Task ProcessDeploymentStepAsync(DeploymentStep step, DeploymentPlanResult result)
    {
        if (step is not TDeploymentStep deploymentStep)
        {
            return Task.CompletedTask;
        }

        return ProcessAsync(deploymentStep, result);
    }
```

Reply:
`ProcessAsync()` has the same signature of `ProcessDeploymentStepAsync()` I'm trying to reduce the boiler plate code as much as we can

Reply:
Yeah but I would not make DeploymentStep as global. I would keep that are the first argument of ProcessAsync method instead.

Reply:
@hishamco it's not the same as AdminMenu. Here you should pass the casted type to the abstract method as a first argument. Int he AdminMenu, you almost never need to access `Name` property so it is okay for that to be stored as a protected property. For the deployment steps, you almost always need the `step`. For that reason, its better to pass that to the method as argument. This will also slightly reduce the changes to your PR.

Reply:
I think it will not make a difference if we expose it as `protected` property, otherwise, no one will see much difference what it was

Reply:
you cannot get this method to work "almost never" unless that step is available to it. So it does not make sense to me to not pass that argument to the function. 

The benefit that you are adding here is cleaning up code without introducing a new behavior that you'll have to explain. `ProcessAsync` is already a nicer name. + we no longer need to case the generic step to a concrete type. These are the only two advantages that I am expecting here.

Reply:
> you cannot get this method to work "almost never" unless that step is available to it. So it does not make sense to me to not pass that argument to the function.

How it supposed to work if the step is not available, could you please elborate

Reply:
I'm not sure when is it needed to run if the step is not available?

Reply:
> you cannot get this method to work "almost never" unless that step is available to it.

I am saying this method is useless without the step. Which is why I am suggesting adding `Step` as a first argument to the method.

Reply:
I'm not sure why it's useless, no need to have a condition in each deployment during processing

Reply:
I am not saying to do that @hishamco. Look again what I suggested here https://github.com/OrchardCMS/OrchardCore/pull/16842#discussion_r1790632383

Reply:
Your update is quite similar except you removed the use of `DeploymentStep` and passed it in the functions

Reply:
Exactly. That is what have requested like 4 times already :)

Reply:
I'm against that, let us continue in Discord or WhatsApp to understand each other